Yannick261

yvanhuele opened the pull request:
Made changes to deal with Issue #261.

Now, if the model was built (and the data cleaned) through supervised_model_trainer, the pre-dummified columns are used to check for missing variables in the prediction data.  To correctly perform dummification on the prediction data, columns of dtype object are converted to dtype category where the categories are taken from the training data values.  One side effect is that new categories not seen in the training data are set to NaN (and then imputed or dropped) instead of being assigned the base class determined by the dummification (usually the category coming first alphabetically) as before.
- Added information about the training data pre-dummified columns and categorical columns to supervised_model_trainer, advanced_supervised_model_trainer, and trained_supervised_model
- Modified DataFrameImputer to impute with columns of dtype 'category' (issue #105)
- Cast prediction dataframe columns of dtype object to dtype category where the categories were taken from the training set to allow for proper dummification
- Added a message to warn the user if new levels of a categorical variable appear in the prediction data
- Removed requirement that response variable be included in prediction data

I ran the example scripts (regression, classification, and advanced) to verify that they still work.  I built the following toy data sets
```{python}
np.random.seed(5077)

df = pd.DataFrame({"id":range(30),
                   "A":np.random.normal(size=30),
                   "B":np.random.normal(size=30),
                   "C":np.random.choice(["red", "blue", "green"], size=30)},
                  columns=["id", "A","B","C"])

df["dot"] = 5*df["A"] - 6*df["B"]
df.loc[df["C"] == 'red', "dot"] += 4
df.loc[df["C"] == 'blue', "dot"] += -6
df.loc[df["C"] == 'green', "dot"] += 2

df["dot2"] = df["dot"] + np.random.normal(size=30, scale=2)

df["num_response"] = df["dot2"] + np.random.normal(size=30, scale = 1)
df["cat_response"] = 'N'
df.loc[df["dot2"] > 0, "cat_response"] = 'Y'

df_regression = df.drop(["dot", "dot2", "cat_response"], axis=1)
df_classification = df.drop(["dot", "dot2", "num_response"], axis=1)

np.random.seed()
```
and verified the following properties hold, but I would appreciate it if you could double check them (preferably not on the exact same data):

- [ ] doesn't raise error if some categories present in the training data are missing in the prediction data (e.g., able to predict a single row of new data)
- [ ] **does** raise error if entire categorical variable is missing entirely in the prediction data (still checks for missing columns)
- [ ] individual predictions are independent of the other rows in the prediction data (i..e., dummification is independent of the prediction data)
- [ ] performs the desired imputation on new categories not present in the training data

Hanlei-Zhu commented:
I checked the four properties and they hold for my dataset. That's great!

Aylr commented on healthcareai/supervised_model_trainer.py:
```diff
@@ -47,6 +47,26 @@ def __init__(self, dataframe, predicted_column, model_type, impute=True, grain_c
         self._advanced_trainer.train_test_split()
 
+        # Save the original dataframe columns to the parent class
+        self._advanced_trainer.pre_dummified_columns = dataframe.columns.values
```
This is a little tricky for a few reasons.
1. This takes the original dataframe after the pipeline has been run, which drops some columns (DTS, grain columns, etc), and we don't necessarily want to enforce those being there for predictions to work.
2. I would prefer if these original column values are an optional constructor argument so it is more clear to users how to use the class, rather than setting a property after the fact.

yvanhuele replied:
I changed the name of pre_dummified_columns to colnames_pre_pipeline and passed it as an optional argument to the constructor.  The name and documentation should make it clear when to use this argument.

Aylr commented on healthcareai/supervised_model_trainer.py:
```diff
@@ -47,6 +47,26 @@ def __init__(self, dataframe, predicted_column, model_type, impute=True, grain_c
+        # Save the original dataframe columns to the parent class
+        self._advanced_trainer.pre_dummified_columns = dataframe.columns.values
+
+        # Identify the categorical columns
```
Good work here. A few thoughts:
1. I wonder if this should be part of the advanced class, since advanced users will likely run into this problem too.
2. This chunk of code from 53 to 66 might do well as its own function, rather than adding to the complexity of the constructor.

yvanhuele replied:
Re 1: My understanding was that all of the data cleaning -- dummification in particular -- happens before the advanced trainer is ever used.  Is there a good way to move this functionality to the advanced trainer without also pushing all the data cleaning into the advanced trainer?

Aylr commented on healthcareai/trained_models/trained_supervised_model.py:
```diff
@@ -55,6 +57,7 @@ def __init__(self,
             test_set_actual (list): y_test
             metric_by_name (dict): Metrics by name
+            pre_dummified_columns (list): List of column names used as features before dummification
         """
```
Need docstring for `categorical_column_info`

Aylr replied:
Please. :smile:

yvanhuele replied:
Nice catch.  I'll get right to it.

Aylr commented on healthcareai/trained_models/trained_supervised_model.py:
```diff
@@ -183,17 +188,37 @@ def prepare_and_subset(self, dataframe):
             pandas.core.frame.DataFrame: A dataframe that has been run through the pipeline and subsetted to only the columns the model expects.
         """
 
+        # Add response column if not included in prediction data
```
Please clarify comment - add prediction column
...and this (lines 191-193) should probably be it's own function that we could easily test.

yvanhuele replied:
I changed the comment to explain the reasoning behind adding a prediction column full of NaNs  and added a condition so that the if statement only triggers if the prediction column is found in colnames_pre_pipeline.

Let me know if you'd still like me to make this into its own function.

Aylr commented on healthcareai/trained_models/trained_supervised_model.py:
```diff
@@ -183,17 +188,37 @@ def prepare_and_subset(self, dataframe):
+                    # a message warning that these new values will be dropped and imputed
+                    new_values = {v for v in dataframe[column].unique() if not (v in col_categories or pd.isnull(v))}
+                    if len(new_values) > 0:
+                        category_message = 'Column {} contains levels not seen in the training set. These levels have been '
```
Python 3 allows for implied multi-line strings that will clean this up (no need for string concatenation +=)

yvanhuele replied:
Awesome! I'll go ahead and clean that up.

Somewhat related: Do you think the message about new levels being removed would be better as a warning (rather than just a print statement)?

Aylr commented on healthcareai/trained_models/trained_supervised_model.py:
```diff
@@ -183,17 +188,37 @@ def prepare_and_subset(self, dataframe):
             # Subset the dataframe to only columns that were saved from the original model training
             prepared_dataframe = prepared_dataframe[self.column_names]
         except KeyError as ke:
+            # TODO Case for pre_dummified_columns != None
```
You'll have to explain to me what this TODO means...

yvanhuele replied:
The idea behind this was to list different required columns depending on whether or not dummification occurred.  

For example, suppose the training data contained a gender column with values female and male and a color column with values blue, green, and purple, and the test data is missing the gender column and only has blue in the color column.  In this case, I want the error to simply report that the gender column is missing instead of listing gender.male, color.green, and color.purple as missing columns.

I'll go ahead and modify the error message and get rid of the cryptic TODO.

Aylr commented on healthcareai/common/transformers.py:
```diff
@@ -10,7 +10,7 @@ class DataFrameImputer(TransformerMixin):
     Impute missing values in a dataframe.
 
-    Columns of dtype object (assumed categorical) are imputed with the mode (most frequent value in column).
+    Columns of dtype object or category (assumed categorical) are imputed with the mode (most frequent value in column).
```
Nice

Aylr commented on healthcareai/supervised_model_trainer.py:
```diff
@@ -191,6 +198,31 @@ def advanced_features(self):
         """ Returns the underlying AdvancedSupervisedModelTrainer instance. For advanced users only. """
         return self._advanced_trainer
 
+    def categorical_column_info(self, dataframe):
```
This looks good and needs a unit test.

Aylr commented on healthcareai/supervised_model_trainer.py:
```diff
@@ -191,6 +198,31 @@ def advanced_features(self):
+            value_distribution = dataframe[column].value_counts(sort=False)
+            # Sort by the index to ensure the correct dummy is dropped in get_dummies(drop_first=True)
+            value_distribution.sort_index(inplace=True)
+            value_distribution *= 1 / value_distribution.values.sum()
```
rewrite this line in the more verbose (but easier to understand) way

yvanhuele commented:
I moved the function get_categorical_levels() to its own file in healthcareai/common and added unit tests to 
- test that single row predictions work
- test that predictions are independent of the other rows in the dataframe
- test that an error is raised if a column is fully missing (numeric or categorical)
- test that new category levels are imputed correctly
- test that get_categorical_levels() behaves as expected
